server/src/unifyfs_service_manager.c

```diff
@@ -102,141 +104,201 @@ int sm_issue_chunk_reads(int src_rank,
+             * request is for the same app_id */
             last_app = app_id;
         }
-        int spillfd = app_config->spill_log_fds[cli_id];
-        char* log_ptr = app_config->shm_superblocks[cli_id] +
-                        app_config->data_offset + offset;
 
-        char* buf_ptr = databuf + buf_cursor;
+        /* client id for this read task */
+        int cli_id = rreq->log_client_id;
+
+        /* get size of data region for this
+         * app_id and client_id */
+        size_t data_size = app_config->data_size;
 
         /* prepare read opertions based on data location */
-        size_t sz_from_mem = 0;
+        size_t sz_from_mem   = 0;
         size_t sz_from_spill = 0;
-        if ((offset + size) <= app_config->data_size) {
+        if ((offset + size) <= data_size) {
             /* requested data is totally in shared memory */
             sz_from_mem = size;
-        } else if (offset < app_config->data_size) {
+        } else if (offset < data_size) {
             /* part of the requested data is in shared memory */
-            sz_from_mem = app_config->data_size - offset;
+            sz_from_mem   = data_size - offset;
             sz_from_spill = size - sz_from_mem;
         } else {
             /* all requested data is in spillover file */
             sz_from_spill = size;
         }
+
+        /* get pointer to next position in buffer to store read data */
+        char* buf_ptr = databuf + buf_cursor;
+
+        /* read data from shared memory */
         if (sz_from_mem > 0) {
-            /* read data from shared memory */
-            memcpy(buf_ptr, log_ptr, sz_from_mem);
+            /* start of data within in superblock */
+            char* super_addr = app_config->shm_superblocks[cli_id];
+            char* data_addr  = super_addr + app_config->data_offset;
+            char* data_ptr   = data_addr + offset;
+
+            /* copy data from superblock into read reply buffer */
+            memcpy(buf_ptr, data_ptr, sz_from_mem);
+
+            /* we assume memcpy copied everything */
             rresp->read_rc = sz_from_mem;
         }
+
+        /* read data from spillover file */
         if (sz_from_spill > 0) {
-            /* read data from spillover file */
-            ssize_t nread = pread(spillfd, (buf_ptr + sz_from_mem),
-                                  sz_from_spill, 0);
+            /* file descriptor for open spillover file for this
+             * app/client */
+            int spill_fd = app_config->spill_log_fds[cli_id];
+
+            /* offset within spill over file, need to subtract off
+             * range of offsets that land in data region of
+             * superblock */
+            off_t spill_offset = (off_t)(offset - data_size +
+                sz_from_mem);
+
+            /* read data from the spillover file */
+            ssize_t nread = pread(spill_fd, (buf_ptr + sz_from_mem),
+                                  sz_from_spill, spill_offset);
```

Comment:
The fix is in the pread change, where is was using 0 as the last argument before (old line 201), it's now using an offset into the file based on the log_offset.  I had that singular change as its own commit and then layered a bunch of style edits and new comments on top of that in a later commit, but ended up merging these all into one to get past a checkpatch complaint.